include/unifex/detail/debug_async_scope.hpp

```diff
@@ -0,0 +1,209 @@
+
+  const std::type_info& concreteType;
+  const char* demangled;
+
+  static const char* demangle(const std::type_info& concreteType) noexcept {
+    int status = -1;
+    auto result = abi::__cxa_demangle(concreteType.name(), nullptr, 0, &status);
+    UNIFEX_ASSERT(status == 0);
+    return result;
+  }
+
+  op_base* next{nullptr};
+  op_base* prev{nullptr};
+};
+
+using debug_ops_t = intrusive_list<op_base, &op_base::next, &op_base::prev>;
+
+struct debug_op_list final {
+  std::mutex mutex_;
+  debug_ops_t ops_;
+
+  void register_debug_operation(op_base* op) noexcept {
+    std::lock_guard lock{mutex_};
+    ops_.push_back(op);
+  }
+
+  void deregister_debug_operation(op_base* op) noexcept {
+    std::lock_guard lock{mutex_};
+    ops_.remove(op);
+  }
+};
+
+template <typename Receiver>
+struct _operation final {
+  struct type;
+};
+
+template <typename Receiver>
+struct _operation<Receiver>::type : op_base {
+  template <typename Receiver2>
+  type(
+      const std::type_info& t,
+      debug_op_list* ops,
+      Receiver2&& receiver) noexcept  // TODO
+    : op_base{t}
+    , ops_(ops)
+    , receiver_(static_cast<Receiver2&&>(receiver)) {}
+
+  type(type&&) = delete;
+
+  debug_op_list* ops_;
+  UNIFEX_NO_UNIQUE_ADDRESS Receiver receiver_;
+
+  template <typename Func>
+  void complete(Func func) noexcept {
+    ops_->deregister_debug_operation(this);
+    func(std::move(receiver_));
+  }
+};
+
+template <typename Receiver>
+struct _receiver final {
+  struct type;
+};
+
+template <typename Receiver>
+struct _receiver<Receiver>::type final {
+  typename _operation<Receiver>::type* op_;
+
+  template <typename... Values>
+  void set_value(Values&&... values) noexcept {
+    op_->complete([&](Receiver&& receiver) noexcept {
+      UNIFEX_TRY {
+        unifex::set_value(
+            std::move(receiver), static_cast<Values&&>(values)...);
+      }
+      UNIFEX_CATCH(...) {
+        unifex::set_error(std::move(receiver), std::current_exception());
+      }
+    });
+  }
+
+  template <typename E>
+  void set_error(E&& e) noexcept {
+    op_->complete([&e](Receiver&& receiver) noexcept {
+      unifex::set_error(std::move(receiver), static_cast<E&&>(e));
+    });
+  }
+
+  void set_done() noexcept { op_->complete(unifex::set_done); }
+};
+
+template <typename Sender, typename Receiver>
+struct _operation_impl final {
+  struct type;
+};
+
+template <typename Sender, typename Receiver>
+using debug_op = typename _operation_impl<Sender, Receiver>::type;
+
+template <typename Sender, typename Receiver>
+struct _operation_impl<Sender, Receiver>::type final
+  : _operation<Receiver>::type {
+  using base_op_t = typename _operation<Receiver>::type;
+  using receiver_t = typename _receiver<Receiver>::type;
+
+  template <typename Sender2, typename Receiver2>
+  type(
+      debug_op_list* ops,
+      Sender2&& sender,
+      Receiver2&& receiver) noexcept  // TODO
+    : base_op_t(typeid(type), ops, static_cast<Receiver2&&>(receiver))
```

Comment:
`typeid(type)` is always going to include this wrapper operation state's type, which is noise, right?
```suggestion
    : base_op_t(typeid(op_t), ops, static_cast<Receiver2&&>(receiver))
```

Reply:
Or maybe even this, which would give you the name of the type that _would_ be the operation state if not for the debug wrapper.
```suggestion
    : base_op_t(typeid(connect_result_t<Sender, Receiver>), ops, static_cast<Receiver2&&>(receiver))
```